### PR TITLE
Ksql 660 schema registry clean up

### LINKED ISSUE #660: Properly handle avro schemas for internal topics in the Schema Registry
When avro format is used schema for internal topics also is added to the schema registry. 
We should either not add these schemas to the schema registry or if we add them we should clean them up when query is terminated.

### PR DESCRIPTION
**Pull request by hjafarpour**
When using schema registry for avro format we need to clean up the schemas for the internal topics when a query is terminated. Also we need to delete the schema for dropped stream/tables.
This PR fixes #660 .

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 1]
```diff
@@ -481,37 +496,20 @@ public DDLCommandExec getDDLCommandExec() {
...
       return false;
     }
-    livePersistentQueries.remove(queryMetadata);
-    allLiveQueries.remove(queryMetadata);
+    livePersistentQueries.remove(persistentQueryMetadata);
+    allLiveQueries.remove(persistentQueryMetadata);
     if (closeStreams) {
-      queryMetadata.close();
+      persistentQueryMetadata.close();
+      new StreamsTopologyUtil()
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: This is a utility class, so should not be instantiated.

File: ksql-engine/src/main/java/io/confluent/ksql/util/StreamsTopologyUtil.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+public class StreamsTopologyUtil {
+
+  private Set<String> getInternalTopicSet(
+      QueryMetadata queryMetadata,
+      SchemaRegistryClient schemaRegistryClient) throws IOException, RestClientException {
+
+    Set<String> internalTopicNames = new HashSet<>();
+    Collection<String> allSubjects = schemaRegistryClient.getAllSubjects();
+
+    for (String subjectName: allSubjects) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/StreamsTopologyUtil.java]: Using streaming API for cleaner code, e.g.

```java
final String suffix1 = KsqlConstants.STREAMS_CHANGELOG_TOPIC_SUFFIX
                          + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX;
final String suffix2 = KsqlConstants.STREAMS_REPARTITION_TOPIC_SUFFIX
                           + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX;
    
return schemaRegistryClient.getAllSubjects().stream()
  .filter(subjectName -> subjectName.startsWith(queryMetadata.getQueryApplicationId()))
  .filter(subjectName -> subjectName.endsWith(suffix1) || subjectName.endsWith(suffix2))
  .collect(Collectors.toSet());
```

File: ksql-engine/src/main/java/io/confluent/ksql/util/StreamsTopologyUtil.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException;
+
+
+public class StreamsTopologyUtil {
+
+  private Set<String> getInternalTopicSet(
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/StreamsTopologyUtil.java]: `private` methods below `public`?

File: ksql-engine/src/main/java/io/confluent/ksql/util/StreamsTopologyUtil.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  public void cleanUpInternalTopicAvroSchemas(QueryMetadata queryMetadata,
+                                              SchemaRegistryClient schemaRegistryClient) {
+    try {
+      Set<String> internalToicSubjects = getInternalTopicSet(queryMetadata, schemaRegistryClient);
+      for (String internalTopicSubject: internalToicSubjects) {
+        schemaRegistryClient.deleteSubject(internalTopicSubject);
+      }
+    } catch (Exception e) {
+      throw new KsqlException("Could not clean up the schema registry for internal topics for "
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/StreamsTopologyUtil.java]: Might be useful to know check schema couldn't be deleted.

File: ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java [hunk 5]
```diff
@@ -124,4 +127,23 @@ public void shouldNotEnforceTopicExistanceWhileParsing() throws Exception {
 
   }
 
+  @Test
+  public void shouldCleanUpSchemaRegistry() throws Exception {
+    final List<QueryMetadata> queries
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java]: Be nice to have each statement on its own line, plus `queries` is never used.

```java
ksqlEngine.buildMultipleQueries(
        "create table bar with (value_format = 'avro') as select * from test2;" +
        "create table foo as select * from test2;",
        Collections.emptyMap());
```

File: ksql-engine/src/main/java/io/confluent/ksql/util/StreamsTopologyUtil.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException;
+
+
+public class StreamsTopologyUtil {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/StreamsTopologyUtil.java]: Can we make this stateless util class a proper util class please?

Util class should:
- be final
- have a private constructor
- only have static methods

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 7]
```diff
@@ -82,8 +82,9 @@
 import io.confluent.ksql.util.Pair;
 import io.confluent.ksql.util.PersistentQueryMetadata;
 import io.confluent.ksql.util.QueryMetadata;
+import io.confluent.ksql.util.StreamsTopologyUtil;
 
-public class KsqlEngine implements Closeable, QueryTerminator {
+public class KsqlEngine implements Closeable {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: You've removed the implementation of `QueryTerminator` but you've left the unused interface around. Futher, this changes the behaviour of the `DropStream` and `DropTable` commands in that any queries against them are not dropped. Either this is not valid or i think these commands should check that all queries have been terminated before a corresponding `DropXxxx` Command is run

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java (hunk 1 above)
Comment (dguy) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: Or maybe we don't need it. i.e., the set of internal topics belongs to the query. So maybe the `QueryMetadata` could have a method something like `internalTopicNames(final Collection<String> topics /*this is the data from schemaRegistryClient.getAllSubjects()*/`
The `cleanupInternalTopicAvroSchemas()` method becomes part of `KsqlEngine` and we delete the util class

File: ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java [hunk 8]
```diff
@@ -124,4 +127,23 @@ public void shouldNotEnforceTopicExistanceWhileParsing() throws Exception {
...
+                                   + "test2;" +
+                                   "create table foo as select * from test2;",
+                                          Collections.emptyMap());
+    Schema schema = SchemaBuilder
+        .record("Test").fields()
+        .name("clientHash").type().fixed("MD5").size(16).noDefault()
+        .endRecord();
+    ksqlEngine.getSchemaRegistryClient().register("BAR-value", schema);
+
+    Assert.assertTrue(schemaRegistryClient.getAllSubjects().contains("BAR-value"));
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java]: `assertThat()` same below

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java [hunk 9]
```diff
@@ -52,7 +52,7 @@
   private final KafkaTopicClient topicClient = EasyMock.createNiceMock(KafkaTopicClient.class);
   private final CommandFactories commandFactories = new CommandFactories(
       topicClient,
-      EasyMock.createMock(QueryTerminator.class),
+      EasyMock.createMock(CachedSchemaRegistryClient.class),
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java]: use the interface, i.e., `SchemaRegistryClient`

**Comment by apurvam** [comment]
I think this should target master. Might be easier to take the changes and reapply them on a fresh checkout of master rather than fixing this PR.

File: ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java [hunk 10]
```diff
@@ -95,6 +102,36 @@ public void close() {
...
+
+
+  public void cleanUpInternalTopicAvroSchemas(SchemaRegistryClient schemaRegistryClient) {
+    try {
+      Set<String> internalToicSubjects = getInternalTopicSet(schemaRegistryClient);
+      for (String internalTopicSubject: internalToicSubjects) {
+        schemaRegistryClient.deleteSubject(internalTopicSubject);
+      }
+    } catch (Exception e) {
+      // Do nothing! Schema registry clean up is best effort!
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java]: Should we at least log a warning?

File: ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java [hunk 11]
```diff
@@ -95,6 +102,36 @@ public void close() {
...
+    return schemaRegistryClient.getAllSubjects().stream()
+        .filter(subjectName -> subjectName.startsWith(getQueryApplicationId()))
+        .filter(subjectName -> subjectName.endsWith(suffix1) || subjectName.endsWith(suffix2))
+        .collect(Collectors.toSet());
+  }
+
+
+  public void cleanUpInternalTopicAvroSchemas(SchemaRegistryClient schemaRegistryClient) {
+    try {
+      Set<String> internalToicSubjects = getInternalTopicSet(schemaRegistryClient);
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java]: maybe

```
getInternalTopicSet(srClient).foreach(srClient::deleteSubject);
```

... embrace Java8 streams! ;)

File: ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java [hunk 12]
```diff
@@ -95,6 +102,36 @@ public void close() {
     }
   }
 
+  private Set<String> getInternalTopicSet(
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java]: Looks to be some duplication here, with the clean up code in `KafkaTopicClientImpl.deleteInternalTopics`, which also needs to get the list of internal topics.   These should probably share an implementation.

File: ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java (hunk 10 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java]: Yes, added a warning.

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 13]
```diff
@@ -278,7 +278,8 @@ public QueryMetadata getQueryExecutionPlan(final Query query) throws Exception {
       final String statementString,
       final MetaStore tempMetaStore,
       final MetaStore tempMetaStoreForParser,
-      final Map<String, Object> overriddenProperties
+      final Map<String, Object> overriddenProperties,
+      final boolean enforceTopicExistence
```
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: Why did we add this parameter? I only ever see false being passed for it.

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 14]
```diff
@@ -307,13 +308,15 @@ public QueryMetadata getQueryExecutionPlan(final Query query) throws Exception {
...
+          tempMetaStoreForParser,
+          false
       );
       ddlCommandExec.tryExecute(
           new RegisterTopicCommand(
               (RegisterTopic) statement
           ),
-          tempMetaStore
+          tempMetaStore,
+          false
```
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: This should be true here right? This is called from the validation phase, before the query makes it onto the command topic.

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 15]
```diff
@@ -343,46 +349,56 @@ public QueryMetadata getQueryExecutionPlan(final Query query) throws Exception {
...
+                                    DataSource.DataSourceType.KSTREAM,
+                                    schemaRegistryClient),
+                                tempMetaStore,
+                                false);
+      ddlCommandExec.tryExecute(new DropSourceCommand(
+                                    (DropStream) statement,
+                                    DataSource.DataSourceType.KSTREAM,
+                                    schemaRegistryClient),
+                                tempMetaStoreForParser,
+                                false);
```
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: This should be true here right? This is called from the validation phase, before the query makes it onto the command topic.

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/DdlCommandExec.java [hunk 16]
```diff
@@ -37,26 +37,28 @@ public DdlCommandExec(MetaStore metaStore) {
   /**
    * execute on temp metaStore
    */
-  public DdlCommandResult tryExecute(DdlCommand ddlCommand, MetaStore tempMetaStore) {
+  public DdlCommandResult tryExecute(DdlCommand ddlCommand, MetaStore tempMetaStore,
```
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/DdlCommandExec.java]: Should isValidatePhase ever be false here? We would only run against a temp meta store if we are validating right? If that's the case we should get rid of that parameter. That would address my concern with buildSingleQueryAst

File: ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java (hunk 5 above)
Comment (rodesai) [reply ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java]: We should still get rid of the `final List<QueryMetadata> queries = `

File: ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java [hunk 17]
```diff
@@ -124,4 +127,45 @@ public void shouldNotEnforceTopicExistanceWhileParsing() throws Exception {
...
+            "create table bar with (value_format = 'avro') as select * from test2;"
+            + "create table foo as select * from test2;",
+            Collections.emptyMap());
+    Schema schema = SchemaBuilder
+        .record("Test").fields()
+        .name("clientHash").type().fixed("MD5").size(16).noDefault()
+        .endRecord();
+    ksqlEngine.getSchemaRegistryClient().register("BAR-value", schema);
+
+    assertThat(schemaRegistryClient.getAllSubjects().contains("BAR-value"), equalTo(true));
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java]: Could write as `assertThat(schemaRegistryClient.getAllSubjects(), hasItem("BAR-value"));`

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java (hunk 14 above)
Comment (rodesai) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: Ignore this copy of the comment. I mean't to write it for line 379

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java (hunk 13 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: Yes, we don't need it here. Removed it.

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java (hunk 15 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: The validation for Drop statements is done in the `KsqlResource.java` class.

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/DdlCommandExec.java (hunk 16 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/DdlCommandExec.java]: Yes, it will always bee false. I removed this.

**Pull request merged**